<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Pyblast by a-slide</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Pyblast</h1>
      <h2 class="project-tagline">Simple and lightweight Python 2.7 wrapper module for NCBI BLAST+</h2>
      <a href="https://github.com/a-slide/pyBlast" class="btn">View on GitHub</a>
      <a href="https://github.com/a-slide/pyBlast/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/a-slide/pyBlast/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="pyblast-01" class="anchor" href="#pyblast-01" aria-hidden="true"><span class="octicon octicon-link"></span></a>pyBlast 0.1</h1>

<p><strong>Simple and lightweight Python 2.7 wrapper module for NCBI BLAST+</strong></p>

<p><strong>Creation : 2015/05/18</strong></p>

<p><strong>Last update : 2015/05/22</strong></p>

<h2>
<a id="blasthit" class="anchor" href="#blasthit" aria-hidden="true"><span class="octicon octicon-link"></span></a>BlastHit</h2>

<p>Python object representing a hit found by blastn. The object contains the following public fields:</p>

<ul>
<li>id: Auto incremented unique identifier [INT]</li>
<li>q_id: Query sequence name [STR]</li>
<li>s_id: Subject sequence name [STR]</li>
<li>identity: % of identity in the hit [FLOAT 0:100]</li>
<li>length: length of the hit [INT &gt;=0]</li>
<li>mis: Number of mismatch in the hit [INT &gt;=0]</li>
<li>gap: Number of gap in the hit [INT &gt;=0]</li>
<li>q_start: Hit start position of the query sequence [INT &gt;=0]</li>
<li>q_end: Hit end position of the query sequence [INT &gt;=0]</li>
<li>s_start: Hit start position of the subject sequence [INT &gt;=0]</li>
<li>s_end: Hit end position of the subject sequence [INT &gt;=0]</li>
<li>evalue: E value of the alignment [FLOAT &gt;=0]</li>
<li>bscore: Bit score of the alignment[FLOAT &gt;=0]</li>
<li>q_seq: Sequence of the query aligned on the subject sequence [STR]</li>
<li>q_orient: Orientation of the query sequence [+ or -]</li>
<li>s_orient: Orientation of the subject sequence [+ or -]</li>
</ul>

<p>The validity of numeric value is checked upon instantiation. Invalid values will raise assertion errors.</p>

<h2>
<a id="blastn" class="anchor" href="#blastn" aria-hidden="true"><span class="octicon octicon-link"></span></a>Blastn</h2>

<p>This class contain the wrapper for Blastn and require the installation of ncbi Blast+ 2.2.28+.</p>

<h3>
<a id="setup-blastn-object-create-subject-database" class="anchor" href="#setup-blastn-object-create-subject-database" aria-hidden="true"><span class="octicon octicon-link"></span></a>Setup Blastn object: Create subject database</h3>

<p>Upon instantiation, a database is created from the user-provided subject sequence. Database files are created in a temporary directory.
The following parameters can be customized at Blastn objects instantiation</p>

<ul>
<li>ref_path: Path to the reference fasta file (not gzipped). Mandatory</li>
<li>makeblastdb_exec: Path of the makeblastdb executable. Default = "makeblastdb"</li>
<li>makeblastdb_opt: makeblastdb command line options as a string. Default = ""</li>
<li>dbtype: Molecule type of target db ('nucl', 'prot'). Default = "nucl"</li>
<li>input_type: type of the data specified in input_file ('asn1_bin', 'asn1_txt', 'blastdb', 'fasta'). Default = "fasta"</li>
</ul>

<p>To ensure a proper database files deletion at the end of the execution it is possible to call the object using the <code>with</code> statement.
Alternatively you can call the <code>rm_db</code> method at the end of the Blastn usage.</p>

<p><strong>Code</strong></p>

<pre><code>with Blastn(ref_path="./subject.fa") as blastn:
    print (blastn)
</code></pre>

<p><strong>Output</strong></p>

<pre><code>CREATE DATABASE: makeblastdb  -dbtype nucl -input_type fasta -in subject.fa -out /tmp/tmpihszgZ/subject

MAKEBLASTDB CLASS   Parameters list
    db_basename subject
    db_dir  /tmp/tmpihszgZ
    db_path /tmp/tmpihszgZ/subject
    dbtype  nucl
    input_type  fasta
    makeblastdb_exec    makeblastdb
    makeblastdb_opt 
    ref_path    subject.fa

Cleaning up blast DB files for "subject"
</code></pre>

<h3>
<a id="calling-blastn-object-perform-blastn-and-return-a-list-of-hits" class="anchor" href="#calling-blastn-object-perform-blastn-and-return-a-list-of-hits" aria-hidden="true"><span class="octicon octicon-link"></span></a>Calling Blastn object: Perform Blastn and return a list of hits</h3>

<p>The Blastn object can then be directly called, as many time as desired, with query fasta files, that can contain several sequences.
The method will automatically call blastn in a multiprocessing fashion, using as many threads as possible.
The following parameters can be customized at Blastn objects calling:</p>

<ul>
<li>query_path: Path to a fasta file containing the query sequences (not gzipped). Mandatory</li>
<li>blast_exec: Path of the blast executable. By Default blastn will be used. Default = "blastn"</li>
<li>blastn_opt: Blastn command line options as a string. Default = ""</li>
<li>task: Type of blast to be performed ('blastn' 'blastn-short' 'dc-megablast' 'megablast' 'rmblastn'). Default = "dc-megablast"</li>
<li>evalue: E Value cuttoff to retain alignments. Default = 1</li>
<li>best_query_hit: find and return only the best hit per query. Default = False</li>
</ul>

<p>A list containing 1 BlastHit object for each query hit found in the subject will be returned, except if not hit were found in which situation 'None' will be returned. 
If the best_query_hit flag was set to True, Only the best hit per query sequence from the query file will be returned.</p>

<p><strong>Code</strong></p>

<pre><code>with Blastn(ref_path="./subject.fa") as blastn:
    hit_list = blastn(query_path="./query.fa")
    for hit in hit_list:
        print (hit)
</code></pre>

<p><strong>Output</strong></p>

<pre><code>CREATE DATABASE: makeblastdb  -dbtype nucl -input_type fasta -in ./subject.fa -out /tmp/tmp1ZBlfT/subject

MAKE BLAST: blastn  -num_threads 4 -task dc-megablast -evalue 1 -outfmt "6 std qseq" -dust no -query ./query.fa -db /tmp/tmp1ZBlfT/subject

    2 hits found
HIT 0   Query   query1:0-48(+)
    Subject subject:19-67(+)
    Lenght : 48 Identity : 100.0%   Evalue : 2e-23  Bit score : 87.8
    Aligned query seq : GCATGCTCGATCAGTAGCTCTCAGTACGCATACGCTAGCATCACGACT

HIT 1   Query   query2:0-48(+)
    Subject subject:89-137(+)
    Lenght : 48 Identity : 100.0%   Evalue : 2e-23  Bit score : 87.8
    Aligned query seq : CGCATCGACTCGATCTGATCAGCTCACAGTCAGCATCAGCTACGATCA

Cleaning up blast DB files for "subject"
</code></pre>

<h2>
<a id="testing-pyblast-module" class="anchor" href="#testing-pyblast-module" aria-hidden="true"><span class="octicon octicon-link"></span></a>Testing pyBlast module</h2>

<p>The module can be easily tested thanks to pytest</p>

<ul>
<li>Install pytest with pip <code>pip instal pytest</code>
</li>
<li>Run test with py.test-2.7  -v</li>
</ul>

<p>Example of output if successful. Please note than some tests might fail due to the random sampling of DNA sequences, and uncertainties of Blastn algorithm.</p>

<pre><code>========================================== test session starts ===========================================
platform linux2 -- Python 2.7.5 -- py-1.4.27 -- pytest-2.7.0 -- /usr/bin/python
rootdir: /home/adrien/Programming/Python/pyBlast, inifile: 
collected 21 items 

test_pyBlast.py::test_BlastHit[4.16866907958-57-98-69-88-12-100-43-1.40452897105-47.3666242716] PASSED
test_pyBlast.py::test_BlastHit[-1-7-10-20-73-54-25-45-98.7921480151-45.2397166228] xfail
test_pyBlast.py::test_BlastHit[8.92741377413--1-100-36-34-33-14-71-18.8547135761-97.6604693294] xfail
test_pyBlast.py::test_BlastHit[10.5987790458-46--1-45-78-81-86-86-73.8740266727-56.887410005] xfail
test_pyBlast.py::test_BlastHit[66.8213911219-62-48--1-91-10-60-20-88.7850139735-81.7901609219] xfail
test_pyBlast.py::test_BlastHit[86.6626174287-29-83-34--1-53-57-68-17.9799756069-7.83036609495] xfail
test_pyBlast.py::test_BlastHit[5.23985331666-43-85-33-7--1-14-3-74.2130782704-88.9289495285] xfail
test_pyBlast.py::test_BlastHit[75.6935977321-8-78-68-10-39--1-74-44.1447867052-22.5203082483] xfail
test_pyBlast.py::test_BlastHit[39.8692596061-60-5-49-77-9-31--1-2.59963139531-46.3133849683] xfail
test_pyBlast.py::test_BlastHit[15.7192632366-24-92-1-64-82-83-90--1-75.5540618409] xfail
test_pyBlast.py::test_BlastHit[18.6627439886-34-57-60-5-45-26-40-77.7840842678--1] xfail
test_pyBlast.py::test_Blastn[blastn-Queries from Subject] PASSED
test_pyBlast.py::test_Blastn[blastn-Random queries] xfail
test_pyBlast.py::test_Blastn[blastn-short-Queries from Subject] PASSED
test_pyBlast.py::test_Blastn[blastn-short-Random queries] xfail
test_pyBlast.py::test_Blastn[dc-megablast-Queries from Subject] PASSED
test_pyBlast.py::test_Blastn[dc-megablast-Random queries] xfail
test_pyBlast.py::test_Blastn[megablast-Queries from Subject] PASSED
test_pyBlast.py::test_Blastn[megablast-Random queries] xfail
test_pyBlast.py::test_Blastn[rmblastn-Queries from Subject] PASSED
test_pyBlast.py::test_Blastn[rmblastn-Random queries] xfail

================================== 6 passed, 15 xfailed in 5.91 seconds ==================================
</code></pre>

<h2>
<a id="dependencies" class="anchor" href="#dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependencies</h2>

<ul>
<li><a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">ncbi Blast+ 2.2.28+</a></li>
<li>
<a href="http://pytest.org/latest/">python package pytest</a>: <code>pip instal pytest</code>
</li>
</ul>

<h2>
<a id="authors-and-contact" class="anchor" href="#authors-and-contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contact</h2>

<p>Adrien Leger - 2015</p>

<ul>
<li>
<a href="mailto:adrien.leger@gmail.com">adrien.leger@gmail.com</a> - <a href="mailto:adrien.leger@inserm.fr">adrien.leger@inserm.fr</a> - <a href="mailto:adrien.leger@univ-nantes.fr">adrien.leger@univ-nantes.fr</a>
</li>
<li><a href="https://github.com/a-slide">Github</a></li>
<li><a href="http://www.atlantic-gene-therapies.fr/">Atlantic Gene Therapies - INSERM 1089</a></li>
</ul>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/a-slide/pyBlast">Pyblast</a> is maintained by <a href="https://github.com/a-slide">a-slide</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>

